Source/layered-grammar.lisp

```diff
@@ -1,8 +1,108 @@
 (in-package #:eu.turtleware.polyclot.impl)
 
-(define-class <aesthetic> () ())
-(define-class <stat>      () ())
-(define-class <geom>      () ())
-(define-class <mods>      () ())
-(define-class <scale>     () ())
-(define-class <coord>     () ())
+(define-class <aest> ()
+  (;; mapping
+   (aest :initarg :aest)
+   (vars :initarg :vars)
+   ;; slots used for caching
+   (last-df :initform nil)
+   (indexes :initform nil)))
+
+(define-class <stat> ()
+  ((aest :initarg :aest)))
+
+;;; <mods> when used in a <stat> context is applied to all rows.
+(define-class <mods>     (<stat>) ())
+(define-class <identity> (<mods>) ())
+
+(define-class <geom> ()
+  ((aest :initarg :aest)))
+
+;;; FIXME stubs for now
+(define-class <scale> () ())
+(define-class <coord> () ())
+(define-class <facet> () ())
+
+(define-class <layer> ()
+  ((data :initarg :data)
+   (aest :initarg :aest)
+   (stat :initarg :stat)
+   (geom :initarg :geom)
+   (mods :initarg :mods)))
+
+(define-class <chart> (<layer>)
+  ((coord :initarg :coord)
+   (scale :initarg :scale)
+   (facet :initarg :facet)
+   (layer :initarg :layer)))
+
+(defun map-aesthetics (aest df row)
+  (with-slots (aest vars last-df indexes) aest
+    (unless (eql last-df df)
+      (setf last-df df
+            indexes (map 'list
+                         (lambda (col)
+                           (position col (cols df) :test #'string=))
+                         vars)))
+    (loop for aes in aest
+          for ind in indexes
+          collect aes
+          collect (ref df row ind))))
+
+;;; CAUTION modifies a dataframe row
+(defun unmap-aesthetics! (aest df row vals)
+  (with-slots (aest vars last-df indexes) aest
+    (unless (eql last-df df)
+      (setf last-df df
+            indexes (map 'list
+                         (lambda (col)
+                           (position col (cols df) :test #'string=))
+                         vars)))
+    (loop for aes in aest
+          for ind in indexes
+          for val = (getf vals aes)
+          do (setf (elt row ind) val))))
+
+(defgeneric collision-modifier (<mods> last vals)
+  (:method ((mods <identity>) last vals)
+    (declare (ignore last))
+    vals))
+
+(defgeneric statistical-transformation (<stat> <data-frame>)
+  (:method ((stat <identity>) (frame <data-frame>))
+    (declare (ignore stat))
+    frame)
+  (:method ((stat <mods>) (frame <data-frame>))
+    (let ((df (copy-data-frame frame))
+          (last-vals nil))
+      (flet ((modify-row (index row)
+               (declare (ignore index))
+               (let* ((aest (aest stat))
```

Comment:
```               (let* ((aest (stat-aest stat))```

Reply:
I've made function aest a dwim operator (as mentioned in the issue thread)